### ISSUE TITLE
some questions about the library: permissions, and try-catch

### ISSUE THREAD

**Issue by AndroidDeveloperLB** [jacktech24/gcmnetworkmanager-android-example#1]
I think you don't need the internet related permissions, and you can use just the boot permission.
Am I correct?
Also, why do you have a try-catch for the scheduling?

**Comment by filipproch** [comment]
@AndroidDeveloperLB you don't need the internet related permissions, they are here just  because I took it from a project where I was using it with some Internet related stuff. About the try-catch, I was getting (using Crashlytics) sometimes crash with scheduling the task, don't know what caused it, but then it's good to try it in a few moments again and it usually works.

**Comment by filipproch** [comment]
@AndroidDeveloperLB permissions removed, close this issue if no more questions

**Comment by AndroidDeveloperLB** [comment]
Didn't Crashlytics tell you what is going on?

**Comment by filipproch** [comment]
@AndroidDeveloperLB it was exception telling nothing, something must have happened on the Play Services side probably, but nothing unusual with modifications some manufacturers are doing on their devices

**Comment by AndroidDeveloperLB** [comment]
A bug in the play services then? Have you reported about this?

**Comment by filipproch** [comment]
@AndroidDeveloperLB I am sure it's not a bug, it was failing on some strange devices (some chinese brand), 99.9% users were okay, and there were other things not working also, but beacause every user matters, I had to fix it this way, you can ignore it and for 99.9% time, you will be okay

**Comment by AndroidDeveloperLB** [comment]
Oh I thought you meant the opposite ("something must have happened on the Play Services side probably" ) .
What did you do in this case, when it failed?

**Comment by filipproch** [comment]
@AndroidDeveloperLB I meant it that way that the error wasn't on my side, probably some implementation of the manufacturer caused it malfunction and throw an exception. My solution was to try it again after some time, which helped because on second or third attempt it finally scheduled the task

**Comment by AndroidDeveloperLB** [comment]
That's too bad. Sounds like you have a failed-task-scheduler-scheduler (a
scheduler for failure of scheduling tasks).  :(

 Do those cases have anything in common ? Other than being some weird
chinese devices?

**Comment by filipproch** [comment]
@AndroidDeveloperLB they were only mobile phones, not rooted, nothing else in common

**Comment by AndroidDeveloperLB** [comment]
Not even Android version? Too bad.
I've noticed you wrote this in code:

```
 //executed between 0 - 10s from now
.setExecutionWindow(30, 60)
```

What is going on here? I've seen a different calculation here  https://developers.google.com/cloud-messaging/network-manager  :

```
long startSecs = 0L; // allow for execution of the task as soon as possible (0s from now)
long endSecs = startSecs + 3600L; // task will be run within 1 min of start time
```

So, if 1 minute is 3600, a single second should be 60 ?!

I think both are wrong, as the docs say it's about seconds:
https://developers.google.com/android/reference/com/google/android/gms/gcm/OneoffTask.Builder#setExecutionWindow(long, long)

I've also noticed that even if I set it to 0,1 , it won't occur right away, and in fact it can run in nearly a minute after being scheduled.

**Comment by filipproch** [comment]
@AndroidDeveloperLB I forgot to change the comment but changed the code, the time there is in seconds. About your notice, yes even if you set it to 0,1, it takes mostly about 10s to execute but as you said, it can be up to minute or so. No way to go around this, maybe it's better to execute the service directly in that case somehow.

**Comment by filipproch** [comment]
@AndroidDeveloperLB added the thing with 0, 1 interval to FAQ

**Comment by AndroidDeveloperLB** [comment]
Well that's another issue to add to the list of issues of this API. :(
The API might be more rich than IOSs and other OSs, but it lacks full-ness...

**Comment by AndroidDeveloperLB** [comment]
Here's the post suggesting more flexibility on this :
https://code.google.com/p/android-developer-preview/issues/detail?id=2713

**Comment by filipproch** [comment]
@AndroidDeveloperLB feel free to send pull request and add link to this issue in README.md if you wan't

**Comment by AndroidDeveloperLB** [comment]
@jacktech24  I've now tried JobScheduler instead of GcmTaskService. It has some differences:
1. It can have the task be executed right away, at least in the POC I've made.
2. It has real bundle as params, and they will get restored even after the device is being restarted.
3. it uses an int id and not a string tag for each task
4. it crashes and tells you to use RECEIVE_BOOT_COMPLETED when you ask the task to be persisted, instead of doing nothing.
5. There is a single builder for tasks, and the API for the service is quite different. Seems a bit "dirty" to me, but it also seems more flexible.
6. I don't think there is the case of cancelling tasks as done on "onInitializeTasks" . There is no similar function there.
7. I think the wakelock values is of 1 minute and not 3, but I also think it's ok to use your own wakelock.

The thing about the time of execution being different is weird, as the GcmTaskService is supposed to use the JobScheduler when runnning on Lollipop.
